gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -2,9 +2,8 @@
 # typed: false
 
 module T::Private::Methods
-  
-  SEMAPHORE = Mutex.new
 
+  @semaphore = Mutex.new
```

Comment:
Heads-up, I won't be surprised if performance overhead of introducing a mutex here will be a non-starter for us. 

I think we'll need to found a solution that's much cheaper. As one option: make sure that replacement is idemporent, rather than synchronized.

Reply:
I'll continue try to investigate a possible fix myself that doesn't involve mutex. 
Appreciate any pointers.

Reply:
+1 to idempotent over a mutex. Basically make this threadsafe. What action is getting pre-empted? Maybe you can hold onto the function instead of re-looking it up or change the order of a few operations?

Reply:
I'll take a look again but I'm still fresh to Ruby so it's taking some time.
Maybe it's missing just a `bind` call since the error I saw was: `UnboundMethod`
https://docs.ruby-lang.org/en/2.2.0/UnboundMethod.html

Which might make sense since the the new method is replaced with another new method.
(which unbinds the previous?)